### ISSUE TITLE
when chaining wrap() calls, on the last takes effect

### ISSUE THREAD

**Issue by jrr** [tgriesser/knex#2899]
# Environment

Knex (and CLI) 0.15.2
Database: `postgres:10.5` docker image

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do

It seems like you can't chain multiple .wrap() calls. For example, if I:

```js
knex.raw("foobar")
  .wrap("sum(",")")
  .wrap("exp(",")")
```

I expect to get `exp(sum(foobar))`, but I seem to get only `exp(foobar)`.

So, the last one wins. Related (?): https://github.com/tgriesser/knex/issues/647

2. Error message

(n/a)

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to mysql or postgresql, then single file example which initializes
   needed data and demonstrates the problem.

(tried runkit but it's currently broken 😐)

**Comment by elhigu** [comment]
I have no idea why wrap method even exists.  It is so much cleaner to write:
```
knex.raw("exp(sum(??))", ['foobar'])
```

> (tried runkit but it's currently broken 😐)

Yeah, I always have to select node 8 for runkit to work 😐

Anyways, as long as `.wrap` is supported, we should either document that it works for only single level or make it work when multiple wrappings are added.

**Comment by mercmobily** [comment]
Yes, the current implementation of the `.wrap()` method doesn't support chaining. Looking at the code in lib/raw.js:62-66:

```javascript
wrap(before, after) {
  this._wrappedBefore = before;
  this._wrappedAfter = after;
  return this;
}
```

Each call to `.wrap()` simply overwrites the previous `_wrappedBefore` and `_wrappedAfter` values, rather than accumulating them.

I agree with @elhigu's comment that using `knex.raw("exp(sum(??))"), ['foobar'])` is indeed cleaner and more explicit. The `.wrap()` method adds complexity to the codebase whilst providing minimal value, especially given that raw SQL already provides a straightforward way to achieve the same result.

However, I see this as a bug and I believe we should fix this issue rather than just document it. The fix is VERY straightforward and maintains backwards compatibility whilst making the API more predictable. Here's what I recommend:

```javascript
wrap(before, after) {
  // Handle null/undefined inputs - treat as empty strings
  before = before == null ? '' : String(before);
  after = after == null ? '' : String(after);
  
  // Accumulate wrappings instead of overwriting to support chaining
  this._wrappedBefore = before + (this._wrappedBefore || '');
  this._wrappedAfter = (this._wrappedAfter || '') + after;
  return this;
}
```
This simple fix:
- Properly chains multiple `.wrap()` calls as users would expect
- Handles edge cases (null, undefined, non-strings)
- Maintains full backwards compatibility for existing single `.wrap()` usage
- Is only a few lines of code change with minimal complexity

I'll implement this fix along with comprehensive tests covering:

- Basic chaining behaviour
- Edge cases (empty strings, null/undefined values)
- Backwards compatibility with single wrap calls

Whilst I agree that using raw SQL directly (e.g., `knex.raw("exp(sum(??))"), ['foobar'])`) is often cleaner, since `.wrap()` exists in the API, it should work as users reasonably expect. We can consider deprecating it in a future major version if we determine it adds unnecessary complexity to the codebase, but for now, a working implementation is better than a broken one.
